src/viam/sdk/services/motion/motion.hpp

```diff
@@ -247,6 +366,63 @@ class Motion : public Service {
         const std::vector<WorldState::transform>& supplemental_transforms,
         const AttributeMap& extra) = 0;
 
+    /// @brief Stop a currently executing motion plan.
+    /// @param component_name the component of the currently executing plan to stop.
+    inline void stop_plan(const Name& component_name) {
+        return stop_plan(component_name, {});
+    }
+
+    /// @brief Stop a currently executing motion plan.
+    /// @param component_name the component of the currently executing plan to stop.
+    /// @param extra Any additional arguments to the method.
+    virtual void stop_plan(const Name& component_name, const AttributeMap& extra) = 0;
+
+    /// @brief Returns the plan and state history of the most recent execution to move a component.
+    /// Returns a result if the last execution is still executing, or changed state within the last
+    /// 24 hours without an intervening robot reinitialization.
+    /// @param component_name The name of the component which the MoveOnGlobe request asked to move.
+    /// @param last_plan_only If true, causes only the last plan for the component/execution to be
+    /// returned. Defaults to false.
+    /// @param execution_id Optional execution id of a previous plan, if you want to know about it.
+    inline std::pair<plan_with_status, std::vector<plan_with_status>> get_plan(
+        Name component_name,
+        bool last_plan_only = false,
+        boost::optional<std::string> execution_id = {}) {
+        return get_plan(component_name, {}, last_plan_only, execution_id);
+    }
+
+    /// @brief Returns the plan and state history of the most recent execution to move a component.
+    /// Returns a result if the last execution is still executing, or changed state within the last
+    /// 24 hours without an intervening robot reinitialization.
+    /// @param component_name The name of the component which the MoveOnGlobe request asked to move.
+    /// @param last_plan_only If true, causes only the last plan for the component/execution to be
+    /// returned.
+    /// @param execution_id Optional execution id of a previous plan, if you want to know about it.
+    /// @param extra Any additional arguments to the method.
+    virtual std::pair<plan_with_status, std::vector<plan_with_status>> get_plan(
+        const Name& component_name,
+        const AttributeMap& extra,
+        bool last_plan_only = false,
+        boost::optional<std::string> execution_id = {}) = 0;
```

Comment:
Note to reviewers: this is a little awkward. `extra` is treated as optional in all methods (hence the `inline` versions) but here we have two other optional methods! I didn't want to require people to pass `execution_id` and/or `last_plan_only` at all times, but the ordering of multiple optional arguments can get a little tricky. Curious if anyone has thoughts on a better API here.

cc @acmorrow who I know has thoughts on defaults args in C++.

Reply:
Yeah, I'm finding the interaction of the virtual methods, the default parameters, and the mandatory handling for `extra` is starting to look pretty bad. I have a few suggestions.

I think the way forward is to admit that there needs to be more than one non-virtual entry point, and then have those delegate to the virtual method, potentially as private per the template method pattern I mentioned before. For instance, rather than having a defaulted bool parameter (more on that later) for whether you only want the last plan, make multiple methods:

`get_plans`
`get_last_plan`
`get_plans_by_id`
`get_last_plan_by_id`

Then you don't need defaults at all, since the signatures can differ (as can the return types!). And then it is easy to tack the optional `extra` on the end. All of these functions would just be wrappers around `get_plan` which would always require all of the parameters.

Meanwhile, regarding defaulted bool parameters, please try to avoid it, because at the call site it becomes very hard to know what it means when specified.